mantidproject/mantid
# ApplyCalibration should handle bar scans

**Description of work.**
- [x] Algorithm accepts variable number of columns.
- [x] Algorithm accepts columns for pixel Y-coordinate, height, and width.
- [x] Unit test involving pixel Y-coordinate, height, and width.
- [x] Update algorithm's documentation, including the doctest

**To test:**
Run the following script which will modify the detector's heights of GPSANS. The script will also generate a workspace named `heights` which contains these modified heights as intentisity values, so that one can visualize them in the instrument view.
The script uses the deprecated property `PositionTable` to verify that a deprecation notice is sent to the Results Log widget.

```
import time
# Resize the detector heights of GPSANS
# We double the height of detectors with even detector ID, and we half the height for the rest
ws = LoadEmptyInstrument(InstrumentName='GPSANS', OutputWorkspace='ws')
compInfo = ws.componentInfo()
detInfo = ws.detectorInfo()

# Create and populate the calibration table
calibTable = CreateEmptyTableWorkspace(OutputWorkspace='CalibTable')
calibTable.addColumn(type='int',name='Detector ID')
calibTable.addColumn(type='double',name='Detector Height')
for i in range(ws.getNumberHistograms()):
    det = ws.getDetector(i)
    detector_id = det.getID()
    compInfo_index = detInfo.indexOf(detector_id)
    if detInfo.isMonitor(compInfo_index) is True:
        continue
    factor = 2 if detector_id % 2 == 0 else 0.5
    height = factor * compInfo.shape(compInfo_index).getBoundingBox().width().Y()
    nextRow = {'Detector ID': detector_id,  'Detector Height': height}
    calibTable.addRow(nextRow)

# Apply the calibration and find out how long it takes
start_time = time.time()
ApplyCalibration(Workspace='ws', PositionTable='CalibTable')
elapsed_time = time.time() - start_time
print('ApplyCalibration took ' + str(int(1000 * elapsed_time)) + ' mili-seconds')

# Visualize the calibration by using the detector heights as the intensities. First we need to collect the new detector heights
ws = mtd['ws']
compInfo = ws.componentInfo()
detInfo = ws.detectorInfo()
heights = list()
for i in range(ws.getNumberHistograms()):
    det = ws.getDetector(i)
    compInfo_index = detInfo.indexOf(det.getID())
    if detInfo.isMonitor(compInfo_index) is True:
        continue
    scaling_y = compInfo.scaleFactor(compInfo_index).Y()
    heights.append(scaling_y * compInfo.shape(compInfo_index).getBoundingBox().width().Y())

# Create a workspace using the heights as intensities
CreateWorkspace(DataX=[0, 1.], DataY=heights, NSpec=len(heights), OutputWorkspace='heights')
LoadInstrument(Workspace='heights', RewriteSpectraMap=True, InstrumentName='GPSANS')
```
Typical output of the script to `STDOUT`:

> Wed Feb 12 14:16:28 2020: Script execution started.
> ApplyCalibration took 82 mili-seconds
> Wed Feb 12 14:16:45 2020: Script execution finished.

And a snapshot of the instrument view of workspace `heights`

<img src="https://user-images.githubusercontent.com/1136006/74369361-3efd0a00-4da3-11ea-9a08-926cd5738d9f.png" width="400">


Fixes #27876



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
